File: 백준/Gold/14698. 전생했더니 슬라임 연구자였던 건에 대하여 （Hard）/전생했더니 슬라임 연구자였던 건에 대하여 （Hard）.py
from queue import PriorityQueue
import sys

input = sys.stdin.readline
print = sys.stdout.write

pq = PriorityQueue()
T = int(input().rstrip())

for i in range(T):
    N = int(input().rstrip())
    result = 1
    while not (pq.empty()):
        pq.get()

    if N == 1:
        a = int(input().rstrip())
        print("1\n")
        continue
    else:
        L = list(map(int, input().split()))
        for j in L:
            pq.put(j)

    while pq.qsize() != 1:
        sum = 0
        sum += pq.get()
        sum *= pq.get()
        pq.put(sum)
        result *= sum

    print(str(result % 1000000007) + "\n")

File: 백준/Gold/14698. 전생했더니 슬라임 연구자였던 건에 대하여 （Hard）/전생했더니 슬라임 연구자였던 건에 대하여 （Hard）.cc
#include <iostream>
#include <queue>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int T, N;
    long long a, sum, result;
    cin >> T;
    priority_queue<long long, vector<long long>, greater<long long>> pq;

    for(int i=0; i<T; i++) {
        cin >> N;
        result=1;
        while(!pq.empty()) pq.pop();

        for(int i=0; i<N; i++) {
            cin >> a;
            pq.push(a);
        }

        if(N==1) {
            cout << 1 << '\n';
            continue;
        }

        while(pq.size()!=1) {
            sum=0;
            sum+=pq.top();
            pq.pop();
            sum*=pq.top();
            pq.pop();
            pq.push(sum);
            result*=(sum%1000000007);
            result%=1000000007;
        }
        cout << result%1000000007 << '\n';
    }
}